### PR TITLE
ALSO-5603 Add parameter hiding autocomplete options from the UI

### PR DESCRIPTION
**Pull request by spodo99**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

The new flag is added to the setting to hide new autocomplete method options, if it will be requested by the users.  Per conversation at https://autodesk.slack.com/archives/C1866AFQB/p1679067914031959

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. **Mandatory section** 

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 1]
```diff
@@ -655,6 +655,11 @@ public struct StartConfiguration
             /// If true, Analytics and Usage options are hidden from UI 
             /// </summary>
             public bool HideReportOptions { get; set; }
+
+            /// <summary>
+            /// If true, autocomplete method options are hidden from UI 
+            /// </summary>
+            public bool HideAutocompleteMethodOptions { get; set; }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:662]: @pinzart @mjkkirschner Is this allowed to add new params in this class?

Comment (pinzart90) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:662]: https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules
looks at `DISALLOWED: Adding an instance field to a struct that has no nonpublic fields`
```
Adding any new fields - public or nonpublic - to such a struct is a source breaking change for these callers, as the compiler will now require the additional fields to be initialized.
```

**Comment by mjkkirschner** [comment]
But doesn’t a bool have a default value ?On Mar 17, 2023, at 3:08 PM, pinzart90 ***@***.***> wrote:﻿
@pinzart90 commented on this pull request.



In src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:
> @@ -655,6 +655,11 @@ public struct StartConfiguration
             /// If true, Analytics and Usage options are hidden from UI 
             /// </summary>
             public bool HideReportOptions { get; set; }
+
+            /// <summary>
+            /// If true, autocomplete method options are hidden from UI 
+            /// </summary>
+            public bool HideAutocompleteMethodOptions { get; set; }

https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules
looks at DISALLOWED: Adding an instance field to a struct that has no nonpublic fields
Adding any new fields - public or nonpublic - to such a struct is a source breaking change for these callers, as the compiler will now require the additional fields to be initialized.

**Comment by pinzart90** [comment]
> But doesn’t a bool have a default value ?

It does, but that is not the issue. Try this code in c#:
```
using System;
				
public struct Test {
	
	public int a; // Was added in version 1.0.0
	public double b; // Was added version 1.0.0
	public bool c;// Was added in version 1.0.1 (breakin change)
}

public class Program
{
	public static void Main()
	{
		Test m;
		
		m.a = 3;
		m.b = 0.4;
		Console.WriteLine(m.c); // Compilation error (line 18, col 21): Use of possibly unassigned field 'c 
	}
}
```

**Comment by pinzart90** [comment]
> @pinzart90 Got it, sorry, in C++ it would fail only if you use initializer list... How would be better to pass this parameter then? Using new argument to DynamoViewModel::Start() function and DynamoViewModel constructor?

I don't know strict we've been in the past about this case. I think we've done this before without knowing it is a breaking  change.
The way we've extended existing classes is to add a new Interface to the Preferences class. There are quite a few now, looke at the interfaces here https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Configuration/IPreferences.cs

**Comment by pinzart90** [comment]
@mjkkirschner any other ideas besides yet another interface ?

**Comment by mjkkirschner** [comment]
> @mjkkirschner any other ideas besides yet another interface ?

no better ideas, except perhaps just modifying the concrete type.

**Comment by mjkkirschner** [comment]
@spodo99 there is a test failure:

[Dynamo.Tests.Configuration.PreferenceSettingsTests.TestImportCopySettings](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/8726/testReport/junit/Dynamo.Tests.Configuration/PreferenceSettingsTests/TestImportCopySettings/)

File: src/DynamoCore/Configuration/IPreferences.cs [hunk 2]
```diff
@@ -182,6 +182,18 @@ internal interface IDisablePackageLoadingPreferences
         bool DisableCustomPackageLocations { get; set; }
     }
 
+    /// <summary>
+    /// Temporary interface to avoid breaking changes.
+    /// TODO: Merge with StartupConfiguration for 3.0 (DYN-1699)
+    /// </summary>
+    internal interface IHideAutocompleteMethodOptions
```
[the remark this answers is not in the document]
Comment (QilongTang) [reply src/DynamoCore/Configuration/IPreferences.cs:189]: @pinzart90 I don't think we do currently, we intend to use Launchdarkly to control beta features and UI aspects in general.

**Comment by QilongTang** [comment]
> @mjkkirschner any other ideas besides yet another interface ?

@pinzart @mjkkirschner I think startup params should also work for them right? See https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoSandbox/DynamoCoreSetup.cs#L40, I guess when Alias launching DynamoSandbox, they can pass in param to modify certain feature sets. Essentially, this might just be similar to the existing way without an additional interface

**Comment by mjkkirschner** [comment]
> Do you mean a startup configuration param? (which is a struct - and had the issue Tibi identified) ? or did you mean a new CLI argument? Or both?

I think that like Tibi said, it's an API break, but it's unclear to me who would be affected.

IMO The new preference is safer, the startup config option seems more idiomatic at the cost of integrators potentially having runtime or compile tie failures. (unlikely I guess).

**Comment by spodo99** [comment]
@mjkkirschner  Added missing option to the new setting file

**Comment by mjkkirschner** [comment]
Hi @spodo99 it looks good to me. I guess ideally it would be good to add a test that asserts the UI is actually hidden? I know UI tests are sometimes painful to write in wpf.

Perhaps because of the time constraints here you could merge this and send a followup for a UI test if it's straightforward enough to write - maybe something like this test:
https://github.com/DynamoDS/Dynamo/blob/master/test/DynamoCoreWpfTests/CoreUITests.cs#L760

**Review by mjkkirschner — APPROVED** [review]
the test failure is a flaky ui test.

**Comment by QilongTang** [comment]
As communicated internally, we will make a 2.17.2 release for this, so this needs to be cherry-picked there

**Comment by spodo99** [comment]
@mjkkirschner , @avidit found that I missed several option controls in the Preferences menu. I created a PR https://github.com/DynamoDS/Dynamo/pull/13859 fixing this. I control visibility of the entire StackPanel containing all ML controls now. Could you please merge it with 2.17.2 as well, if it's ok? 
I will create unit tests later as it was proposed...

**Pull request merged**